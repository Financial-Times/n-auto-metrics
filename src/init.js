let metrics;

export const initAutoMetrics = instance => {
	metrics = instance;
};

export const getMetricsInstance = () => metrics;
