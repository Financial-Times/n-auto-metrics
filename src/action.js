import { getMetricsInstance } from './init';
import { isPromise } from './utils';

const createEventMetrics = ({ service, operation, action }) => {
	const metrics = getMetricsInstance();
	if (metrics === undefined) {
		throw Error(`auto metrics instance needs to be initialised first`);
	}
	const serviceRoot = `service.${service}.action.${action}`;
	const operationRoot = `operation.${operation}.action.${action}`;
	const countOne = path => {
		metrics.count(`${serviceRoot}.${path}`, 1);
		metrics.count(`${operationRoot}.${path}`, 1);
	};
	return {
		start: () => countOne('state.start'),
		success: () => countOne('state.success'),
		failure: e => {
			countOne(`state.failure.category.${e.category}.status.${e.status}`);
			countOne(`state.failure.category.${e.category}.type.${e.type}`);
		},
	};
};

const metricsEvent = ({ service, operation, action }) => {
	const event = createEventMetrics({ service, operation, action });
	event.start();
	return event;
};

export const autoMetricsAction = callFunction => {
	const enhancedFunction = (paramsOrArgs = {}, meta = {}) => {
		const service = meta.service || paramsOrArgs.service || 'undefined';
		const action = callFunction.name;
		if (typeof service !== 'string' || service.includes(' ')) {
			throw Error(
				`action metrics service name needs to be string without spaces, at function ${action}`,
			);
		}
		const operation = meta.operation || paramsOrArgs.operation;

		const event = metricsEvent({ service, operation, action });

		try {
			const call = callFunction(paramsOrArgs, meta);
			if (isPromise(call)) {
				return call
					.then(data => {
						event.success();
						return data;
					})
					.catch(e => {
						event.failure(e);
						throw e;
					});
			}
			const data = call;
			event.success();
			return data;
		} catch (e) {
			event.failure(e);
			throw e;
		}
	};
	Object.defineProperty(enhancedFunction, 'name', {
		value: callFunction.name,
		configurable: true,
	});
	return enhancedFunction;
};

export const autoMetricsActions = service => actionBundle => {
	if (typeof service !== 'string' || service.includes(' ')) {
		throw Error(`service name for metrics needs to be string without spaces`);
	}
	const enhanced = {};
	Object.keys(actionBundle).forEach(methodName => {
		const enhancedMethod = (paramsOrArgs, meta) =>
			autoMetricsAction(actionBundle[methodName])(paramsOrArgs, {
				...meta,
				service,
			});
		Object.defineProperty(enhancedMethod, 'name', {
			value: methodName,
			configurable: true,
		});
		enhanced[methodName] = enhancedMethod;
	});
	return enhanced;
};
