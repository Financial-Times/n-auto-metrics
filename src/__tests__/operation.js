import express from 'express';
import request from 'supertest';
import compose from 'compose-function';

import { initAutoMetrics } from '../init';
import { autoMetricsAction } from '../action';
import {
	toMiddleware,
	toMiddlewares,
	autoMetricsOp,
	autoMetricsOps,
} from '../operation';

const metrics = {
	count: jest.fn(),
};

describe('autoMetricsOp', () => {
	beforeAll(() => {
		initAutoMetrics(metrics);
	});

	afterEach(() => {
		jest.resetAllMocks();
	});

	describe('toMiddleware and autoMetricsOp', () => {
		describe('returns a valid enhanced express middleware', () => {
			it('for non-async middleware', async () => {
				const content = { foo: 'bar' };
				const operation = (meta, req, res) => {
					res.status(200).send(content);
				};
				const middleware = compose(toMiddleware, autoMetricsOp)(operation);
				const app = express();
				app.use('/', middleware);
				const res = await request(app).get('/');
				expect(res.statusCode).toBe(200);
				expect(res.body).toEqual(content);
			});

			it('for async middleware', async () => {
				const content = { foo: 'bar' };
				const operation = async (meta, req, res) => {
					res.status(200).send(content);
				};
				const middleware = compose(toMiddleware, autoMetricsOp)(operation);
				const app = express();
				app.use('/', middleware);
				const res = await request(app).get('/');
				expect(res.statusCode).toBe(200);
				expect(res.body).toEqual(content);
			});

			it('for non-async middleware throwing error', async () => {
				const errorInstance = { status: 404, message: 'Not Found' };
				const operation = (meta, req, res, next) => {
					const e = errorInstance;
					next(e);
					throw e;
				};
				/* eslint-disable no-unused-vars */
				const errorHanlder = (err, req, res, next) => {
					res.status(err.status).send(err);
				};
				/* eslint-enable no-unused-vars */
				const middleware = compose(toMiddleware, autoMetricsOp)(operation);
				const app = express();
				app.use('/', middleware, errorHanlder);
				const res = await request(app).get('/');
				expect(res.statusCode).toBe(404);
				expect(res.body.message).toBe('Not Found');
			});

			it('for async middleware throwing error', async () => {
				const errorInstance = { status: 404, message: 'Not Found' };
				const operation = async (meta, req, res, next) => {
					const e = errorInstance;
					next(e);
					throw e;
				};
				/* eslint-disable no-unused-vars */
				const errorHanlder = (err, req, res, next) => {
					res.status(err.status).send(err);
				};
				/* eslint-enable no-unused-vars */
				const middleware = compose(toMiddleware, autoMetricsOp)(operation);
				const app = express();
				app.use('/', middleware, errorHanlder);
				const res = await request(app).get('/');
				expect(res.statusCode).toBe(404);
				expect(res.body.message).toBe('Not Found');
			});
		});
	});

	describe('logs correctly', () => {
		describe('operation success of', () => {
			it('async function with autoMetricsAction sub action', async () => {
				const callFunction = () => {};
				const operationFunction = async meta => {
					await autoMetricsAction(callFunction)(meta);
				};
				await compose(toMiddleware, autoMetricsOp)(operationFunction)();
				expect(metrics.count.mock.calls).toMatchSnapshot();
			});

			it('non-async function with autoMetricsAction sub action', async () => {
				const callFunction = () => {};
				const operationFunction = meta => {
					autoMetricsAction(callFunction)(meta);
				};
				await compose(toMiddleware, autoMetricsOp)(operationFunction)();
				expect(metrics.count.mock.calls).toMatchSnapshot();
			});

			it('req.metrics.segment from previous middlewares', async () => {
				const metaMiddleware = (req, res, next) => {
					req.metrics = {
						...req.metrics,
						segment: 'MOCK_USER_TYPE',
					};
					next();
				};
				const content = { foo: 'bar' };
				const operationFunction = async (meta, req, res) => {
					res.status(200).send(content);
				};
				const middleware = compose(toMiddleware, autoMetricsOp)(
					operationFunction,
				);
				const app = express();
				app.use('/', metaMiddleware, middleware);
				const res = await request(app).get('/');
				expect(res.statusCode).toBe(200);
				expect(res.body).toEqual(content);
				expect(metrics.count.mock.calls).toMatchSnapshot();
			});
		});

		describe('operation failure of', () => {
			it('non-async function', async () => {
				const errorInstance = {
					status: 500,
					message: 'foo',
					category: 'CUSTOM_ERROR',
					type: 'MOCK_ERROR_TYPE',
				};
				const operation = (meta, req, res, next) => {
					try {
						throw errorInstance;
					} catch (e) {
						next(e);
						throw e;
					}
				};
				const next = jest.fn();
				try {
					await autoMetricsOp(operation)(null, null, null, next);
				} catch (e) {
					expect(e).toBe(errorInstance);
					expect(metrics.count.mock.calls).toMatchSnapshot();
					expect(next.mock.calls).toMatchSnapshot();
				}
			});

			it('async function', async () => {
				const errorInstance = {
					status: 500,
					message: 'foo',
					category: 'CUSTOM_ERROR',
					type: 'MOCK_ERROR_TYPE',
				};
				const operation = async (meta, req, res, next) => {
					try {
						throw errorInstance;
					} catch (e) {
						next(e);
						throw e;
					}
				};
				const next = jest.fn();
				try {
					await autoMetricsOp(operation)(null, null, null, next);
				} catch (e) {
					expect(e).toBe(errorInstance);
					expect(metrics.count.mock.calls).toMatchSnapshot();
					expect(next.mock.calls).toMatchSnapshot();
				}
			});
		});
	});
});

describe('autoMetricsOps and toMiddlewares', () => {
	afterEach(() => {
		jest.resetAllMocks();
	});

	it('enhance and convert each method correctly', async () => {
		const content = { foo: 'bar' };
		const methodA = (meta, req, res) => {
			res.status(200).send(content);
		};
		const methodB = (meta, req, res) => {
			res.status(200).send(content);
		};
		const enhancedController = compose(toMiddlewares, autoMetricsOps)({
			methodA,
			methodB,
		});
		const app = express();
		app.use('/a', enhancedController.methodA);
		app.use('/b', enhancedController.methodB);
		const resA = await request(app).get('/a');
		expect(resA.statusCode).toBe(200);
		expect(resA.body).toEqual(content);
		const resB = await request(app).get('/b');
		expect(resB.statusCode).toBe(200);
		expect(resB.body).toEqual(content);
		expect(metrics.count.mock.calls).toMatchSnapshot();
	});
});
