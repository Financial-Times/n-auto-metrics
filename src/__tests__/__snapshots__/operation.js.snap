// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`autoMetricsOp logs correctly operation failure of async function 1`] = `
Array [
  Array [
    "operation.operation.segment.undefined.state.start",
    1,
  ],
  Array [
    "operation.operation.segment.undefined.state.failure.category.CUSTOM_ERROR.type.MOCK_ERROR_TYPE",
    1,
  ],
]
`;

exports[`autoMetricsOp logs correctly operation failure of async function 2`] = `
Array [
  Array [
    Object {
      "category": "CUSTOM_ERROR",
      "message": "foo",
      "status": 500,
      "type": "MOCK_ERROR_TYPE",
    },
  ],
]
`;

exports[`autoMetricsOp logs correctly operation failure of non-async function 1`] = `
Array [
  Array [
    "operation.operation.segment.undefined.state.start",
    1,
  ],
  Array [
    "operation.operation.segment.undefined.state.failure.category.CUSTOM_ERROR.type.MOCK_ERROR_TYPE",
    1,
  ],
]
`;

exports[`autoMetricsOp logs correctly operation failure of non-async function 2`] = `
Array [
  Array [
    Object {
      "category": "CUSTOM_ERROR",
      "message": "foo",
      "status": 500,
      "type": "MOCK_ERROR_TYPE",
    },
  ],
]
`;

exports[`autoMetricsOp logs correctly operation success of async function with autoMetricsAction sub action 1`] = `
Array [
  Array [
    "operation.operationFunction.segment.undefined.state.start",
    1,
  ],
  Array [
    "service.undefined.action.callFunction.state.start",
    1,
  ],
  Array [
    "operation.operationFunction.action.callFunction.state.start",
    1,
  ],
  Array [
    "service.undefined.action.callFunction.state.success",
    1,
  ],
  Array [
    "operation.operationFunction.action.callFunction.state.success",
    1,
  ],
  Array [
    "operation.operationFunction.segment.undefined.state.success",
    1,
  ],
]
`;

exports[`autoMetricsOp logs correctly operation success of non-async function with autoMetricsAction sub action 1`] = `
Array [
  Array [
    "operation.operationFunction.segment.undefined.state.start",
    1,
  ],
  Array [
    "service.undefined.action.callFunction.state.start",
    1,
  ],
  Array [
    "operation.operationFunction.action.callFunction.state.start",
    1,
  ],
  Array [
    "service.undefined.action.callFunction.state.success",
    1,
  ],
  Array [
    "operation.operationFunction.action.callFunction.state.success",
    1,
  ],
  Array [
    "operation.operationFunction.segment.undefined.state.success",
    1,
  ],
]
`;

exports[`autoMetricsOp logs correctly operation success of req.metrics.segment from previous middlewares 1`] = `
Array [
  Array [
    "operation.operationFunction.segment.MOCK_USER_TYPE.state.start",
    1,
  ],
  Array [
    "operation.operationFunction.segment.MOCK_USER_TYPE.state.success",
    1,
  ],
]
`;

exports[`autoMetricsOps and toMiddlewares enhance and convert each method correctly 1`] = `
Array [
  Array [
    "operation.methodA.segment.undefined.state.start",
    1,
  ],
  Array [
    "operation.methodA.segment.undefined.state.success",
    1,
  ],
  Array [
    "operation.methodB.segment.undefined.state.start",
    1,
  ],
  Array [
    "operation.methodB.segment.undefined.state.success",
    1,
  ],
]
`;
